src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+      const input = inputs[key];
+      if (entry === sb3.BooleanOrSubstackInputStatus) {
+        let blockId: string;
+
+        if (input) {
+          const options = {
+            stage,
+            target,
+            blockData,
+            getBroadcastId,
+            customBlockDataMap,
+            parent: block
+          };
+
+          switch (input.type) {
+            case "blocks":
+              blockId = serializeBlockStack(input.value, options);
+              break;
+            case "block":
+              blockId = serializeBlock(input.value, options);
+              break;
+          }
+        }
+
+        if (blockId) {
+          resultInputs[key] = [BIS.INPUT_BLOCK_NO_SHADOW, blockId];
+        }
+      } else {
+        let valueForShadow;
+        if (input.type === "block") {
+          valueForShadow = initialValues[key];
+          // Special-case some input opcodes for more realistic initial values.
+          switch (entry) {
+            case OpCode.looks_costume:
+              if (target.costumes[0]) {
+                valueForShadow = target.costumes[0].name;
+              }
+              break;
+            case OpCode.sound_sounds_menu:
+              if (target.sounds[0]) {
+                valueForShadow = target.sounds[0].name;
+              }
+              break;
+            case OpCode.event_broadcast_menu:
+              valueForShadow = getBroadcastId.initialBroadcastName;
+              break;
+          }
+        } else {
+          valueForShadow = input.value;
+        }
+
+        const { shadowValue } = serializeInputShadow(valueForShadow, {
+          blockData,
+          getBroadcastId,
+          parentId: block.id,
+          primitiveOrOpCode: entry as number | OpCode
+        });
+
+        let obscuringBlockValue;
+
+        if (input.type === "block") {
+          if (input.value.opcode === OpCode.data_variable) {
+            const variableName = input.value.inputs.VARIABLE.value;
+            const variableId = getVariableId(variableName, target, stage);
+            obscuringBlockValue = [BIS.VAR_PRIMITIVE, variableName, variableId];
+          } else if (input.value.opcode === OpCode.data_listcontents) {
+            const listName = input.value.inputs.LIST.value;
+            const listId = getListId(listName, target, stage);
+            obscuringBlockValue = [BIS.LIST_PRIMITIVE, listName, listId];
+          } else {
+            obscuringBlockValue = serializeBlock(input.value, {
+              blockData,
+              getBroadcastId,
+              customBlockDataMap,
+              parent: block,
+              stage,
+              target
+            });
+          }
+        }
+
+        if (obscuringBlockValue) {
+          resultInputs[key] = [BIS.INPUT_DIFF_BLOCK_SHADOW, obscuringBlockValue, shadowValue];
+        } else {
+          resultInputs[key] = [BIS.INPUT_SAME_BLOCK_SHADOW, shadowValue];
+        }
```

Comment:
```suggestion
        if (input.type === "block") {
          let obscuringBlockValue;

          if (input.value.opcode === OpCode.data_variable) {
            const variableName = input.value.inputs.VARIABLE.value;
            const variableId = getVariableId(variableName, target, stage);
            obscuringBlockValue = [BIS.VAR_PRIMITIVE, variableName, variableId];
          } else if (input.value.opcode === OpCode.data_listcontents) {
            const listName = input.value.inputs.LIST.value;
            const listId = getListId(listName, target, stage);
            obscuringBlockValue = [BIS.LIST_PRIMITIVE, listName, listId];
          } else {
            obscuringBlockValue = serializeBlock(input.value, {
              blockData,
              getBroadcastId,
              customBlockDataMap,
              parent: block,
              stage,
              target
            });
          }

          resultInputs[key] = [BIS.INPUT_DIFF_BLOCK_SHADOW, obscuringBlockValue, shadowValue];
        } else {
          resultInputs[key] = [BIS.INPUT_SAME_BLOCK_SHADOW, shadowValue];
        }
```

Reply:
Nice, thanks! Would a switch/case be better here?

Reply:
Probably